prusti-specs/src/ghost_constraints/mod.rs

```
@@ -0,0 +1,142 @@
+            item_fn.attrs.push(parse_quote_spanned!(tokens_span=>
+                #[prusti::ghost_constraint_trait_bounds_in_where_clause]
+            ));
+        }
+
+        new_items.extend(generated_items);
+        new_attrs.extend(generated_attrs);
+    }
+
+    Ok((new_items, new_attrs))
+}
+
+fn generate_where_clause_for_spec(
+    trait_bounds: &ProvidedTraitBounds,
+    preliminary_where_clause: Option<&syn::WhereClause>,
+) -> syn::WhereClause {
+    let span = trait_bounds.span();
+    if let Some(mut where_clause) = preliminary_where_clause.cloned() {
+        where_clause.predicates.push(parse_quote_spanned!(span=>
+            #trait_bounds
+        ));
+        where_clause
+    } else {
+        parse_quote_spanned!(span=>
+            where #trait_bounds
+        )
+    }
+}
+
+fn verify_provided_trait_bounds(trait_bounds: &ProvidedTraitBounds) -> syn::Result<()> {
+    for bound in &trait_bounds.predicate_type.bounds {
+        match bound {
+            syn::TypeParamBound::Lifetime(lt) => {
+                return Err(syn::Error::new(
+                    lt.span(),
+                    "Lifetimes in ghost constraints not allowed",
+                ))
+            }
+            syn::TypeParamBound::Trait(trait_bound) => {
+                verify_provided_trait_bound(trait_bound)?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn verify_provided_trait_bound(trait_bound: &syn::TraitBound) -> syn::Result<()> {
+    if let Some(lt) = &trait_bound.lifetimes {
+        return Err(syn::Error::new(
+            lt.span(),
+            "Lifetimes in ghost constraints not allowed",
+        ));
+    }
+
+    Ok(())
+}
+
+/// A `syn::Parse`able representaiton of `syn::PredicateType` (ignoring bound lifetimes)
+struct ProvidedTraitBounds {
+    predicate_type: syn::PredicateType,
+}
+
+impl syn::parse::Parse for ProvidedTraitBounds {
```

This parsing logic and the fact that we parse the bounds *at all* is only so that we can raise a parsing error for invalid bounds (e.g. lifetimes)?

---

I think it also comes in handy when generate the where clause for the spec function. Ghost constraint trait bounds are actually `syn::PredicateType`s, but they are not parsable by `syn` (see comment on the type). So I made this custom struct.

Do you think this could be done easier?

---

Well, if we simply passed through the predicate as typed by the user, without parsing it with `syn`, we wouldn't need this. But then some validation (e.g. that lifetime bounds aren't allowed) would need to happen post-typing and maybe that's too much hassle.